File: convert_godot_data.mjs
/**
 * VeilBreakers Godot -> Unity Data Converter
 * Converts .tres (Godot Resource) files to JSON for Unity
 */

import fs from 'fs';
import path from 'path';

const GODOT_PROJECT = 'C:/Users/Conner/Downloads/VeilbreakersGame';
const UNITY_PROJECT = 'C:/Users/Conner/Downloads/VeilBreakers3D';

function parseTresFile(filepath) {
    const content = fs.readFileSync(filepath, 'utf-8');

    // Find the [resource] section
    const resourceMatch = content.match(/\[resource\]([\s\S]*)/);
    if (!resourceMatch) return null;

    const resourceContent = resourceMatch[1];
    const data = {};

    // Parse each line
    const lines = resourceContent.split('\n');
    for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('script')) continue;

        // Match key = value
        const match = trimmed.match(/^(\w+)\s*=\s*(.+)$/);
        if (match) {
            const key = match[1];
            const valueStr = match[2].trim();
            data[key] = parseValue(valueStr);
        }
    }

    return data;
}

function parseValue(valueStr) {
    valueStr = valueStr.trim();

    // Boolean
    if (valueStr === 'true') return true;
    if (valueStr === 'false') return false;

    // null/nil
    if (valueStr === 'null' || valueStr === 'nil') return null;

    // Color(r, g, b, a)
    const colorMatch = valueStr.match(/Color\(\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)\s*\)/);
    if (colorMatch) {
        return {
            r: parseFloat(colorMatch[1]),
            g: parseFloat(colorMatch[2]),
            b: parseFloat(colorMatch[3]),
            a: parseFloat(colorMatch[4])
        };
    }

    // Vector2
    const vec2Match = valueStr.match(/Vector2\(\s*([\d.-]+)\s*,\s*([\d.-]+)\s*\)/);
    if (vec2Match) {
        return { x: parseFloat(vec2Match[1]), y: parseFloat(vec2Match[2]) };
    }

    // String
    if (valueStr.startsWith('"') && valueStr.endsWith('"')) {
        return valueStr.slice(1, -1);
    }

    // Array
    if (valueStr.startsWith('[')) {
        return parseArray(valueStr);
    }

    // Dictionary
    if (valueStr.startsWith('{')) {
        return parseDict(valueStr);
    }

    // Integer
    if (/^-?\d+$/.test(valueStr)) {
        return parseInt(valueStr, 10);
    }

    // Float
    if (/^-?\d+\.\d*$/.test(valueStr)) {
        return parseFloat(valueStr);
    }

    return valueStr;
}

function parseArray(valueStr) {
    try {
        const content = valueStr.slice(1, -1).trim();
        if (!content) return [];

        const items = [];
        let depth = 0;
        let current = '';
        let inString = false;

        for (let i = 0; i < content.length; i++) {
            const char = content[i];

            if (char === '"' && (i === 0 || content[i-1] !== '\\')) {
                inString = !inString;
            }

            if (!inString) {
                if (char === '[' || char === '{') depth++;
                else if (char === ']' || char === '}') depth--;
                else if (char === ',' && depth === 0) {
                    items.push(parseValue(current.trim()));
                    current = '';
                    continue;
                }
            }

            current += char;
        }

        if (current.trim()) {
            items.push(parseValue(current.trim()));
        }

        return items;
    } catch (e) {
        return [];
    }
}

function parseDict(valueStr) {
    try {
        const content = valueStr.slice(1, -1).trim();
        if (!content) return {};

        const result = {};
        let depth = 0;
        let currentKey = null;
        let currentVal = '';
        let inString = false;
        let parsingKey = true;

        for (let i = 0; i < content.length; i++) {
            const char = content[i];

            if (char === '"' && (i === 0 || content[i-1] !== '\\')) {
                inString = !inString;
            }

            if (!inString) {
                if (char === '[' || char === '{') depth++;
                else if (char === ']' || char === '}') depth--;
                else if (char === ':' && depth === 0 && parsingKey) {
                    currentKey = currentVal.trim().replace(/^"|"$/g, '');
                    currentVal = '';
                    parsingKey = false;
                    continue;
                }
                else if (char === ',' && depth === 0) {
                    if (currentKey !== null) {
                        result[currentKey] = parseValue(currentVal.trim());
                    }
                    currentKey = null;
                    currentVal = '';
                    parsingKey = true;
                    continue;
                }
            }

            currentVal += char;
        }

        if (currentKey !== null && currentVal.trim()) {
            result[currentKey] = parseValue(currentVal.trim());
        }

        return result;
    } catch (e) {
        return {};
    }
}

function convertPath(godotPath) {
    if (!godotPath) return '';
    return godotPath.replace('res://', '').replace(/\\/g, '/');
}

function getFilesRecursive(dir, ext) {
    const files = [];
    if (!fs.existsSync(dir)) return files;

    const items = fs.readdirSync(dir);
    for (const item of items) {
        const fullPath = path.join(dir, item);
        const stat = fs.statSync(fullPath);

        if (stat.isDirectory()) {
            files.push(...getFilesRecursive(fullPath, ext));
        } else if (item.endsWith(ext)) {
            files.push(fullPath);
        }
    }

    return files;
}

function convertMonsters() {
    console.log('\n[MONSTERS]');
    const monstersDir = path.join(GODOT_PROJECT, 'data', 'monsters');
    const output = [];

    const files = getFilesRecursive(monstersDir, '.tres');
    for (const file of files) {
        const data = parseTresFile(file);
        if (data) {
            if (data.sprite_path) data.sprite_path = convertPath(data.sprite_path);
            if (data.portrait_path) data.portrait_path = convertPath(data.portrait_path);
            output.push(data);
            console.log(`  Converted: ${data.display_name || path.basename(file)}`);
        }
    }

    return output;
}

function convertSkills() {
    console.log('\n[SKILLS]');
    const skillsDir = path.join(GODOT_PROJECT, 'data', 'skills');
    const output = [];

    const files = getFilesRecursive(skillsDir, '.tres');
    for (const file of files) {
        const data = parseTresFile(file);
        if (data) {
            if (file.includes('monsters')) data.is_monster_skill = true;
            if (file.includes('heroes')) data.is_hero_skill = true;
            if (data.icon_path) data.icon_path = convertPath(data.icon_path);
            output.push(data);
            console.log(`  Converted: ${data.display_name || path.basename(file)}`);
        }
    }

    return output;
}

function convertHeroes() {
    console.log('\n[HEROES]');
    const heroesDir = path.join(GODOT_PROJECT, 'data', 'heroes');
    const output = [];

    const files = getFilesRecursive(heroesDir, '.tres');
    for (const file of files) {
        const data = parseTresFile(file);
        if (data) {
            ['sprite_path', 'portrait_path', 'battle_sprite_path'].forEach(key => {
                if (data[key]) data[key] = convertPath(data[key]);
            });
            output.push(data);
            console.log(`  Converted: ${data.display_name || path.basename(file)}`);
        }
    }

    return output;
}

function convertItems() {
    console.log('\n[ITEMS]');
    const itemsDir = path.join(GODOT_PROJECT, 'data', 'items');
    const output = [];

    for (const subdir of ['consumables', 'equipment']) {
        const subpath = path.join(itemsDir, subdir);
        const files = getFilesRecursive(subpath, '.tres');

        for (const file of files) {
            const data = parseTresFile(file);
            if (data) {
                data.item_category = subdir;
                if (data.icon_path) data.icon_path = convertPath(data.icon_path);
                output.push(data);
                console.log(`  Converted: ${data.display_name || path.basename(file)}`);
            }
        }
    }

    return output;
}

function main() {
    console.log('='.repeat(60));
    console.log('VEILBREAKERS DATA CONVERTER');
    console.log('Godot .tres -> Unity JSON');
    console.log('='.repeat(60));

    // Create output directory
    const outputDir = path.join(UNITY_PROJECT, 'Assets', 'Data');
    fs.mkdirSync(outputDir, { recursive: true });

    // Convert monsters
    const monsters = convertMonsters();
    fs.writeFileSync(path.join(outputDir, 'monsters.json'), JSON.stringify(monsters, null, 2));
    console.log(`  Total: ${monsters.length} monsters\n`);

    // Convert skills
    const skills = convertSkills();
    fs.writeFileSync(path.join(outputDir, 'skills.json'), JSON.stringify(skills, null, 2));
    console.log(`  Total: ${skills.length} skills\n`);

    // Convert heroes
    const heroes = convertHeroes();
    fs.writeFileSync(path.join(outputDir, 'heroes.json'), JSON.stringify(heroes, null, 2));
    console.log(`  Total: ${heroes.length} heroes\n`);

    // Convert items
    const items = convertItems();
    fs.writeFileSync(path.join(outputDir, 'items.json'), JSON.stringify(items, null, 2));
    console.log(`  Total: ${items.length} items\n`);

    console.log('='.repeat(60));
    console.log('CONVERSION COMPLETE!');
    console.log(`Output: ${outputDir}`);
    console.log('='.repeat(60));
}

main();

File: Tools/convert_godot_data.py
#!/usr/bin/env python3
"""
VeilBreakers Godot -> Unity Data Converter
Converts .tres (Godot Resource) files to JSON for Unity
"""

import os
import re
import json
from pathlib import Path

GODOT_PROJECT = Path(r"C:\Users\Conner\Downloads\VeilbreakersGame")
UNITY_PROJECT = Path(r"C:\Users\Conner\Downloads\VeilBreakers3D")

def parse_tres_file(filepath):
    """Parse a Godot .tres file and extract resource data."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # Find the [resource] section
    resource_match = re.search(r'\[resource\](.*)', content, re.DOTALL)
    if not resource_match:
        return None

    resource_content = resource_match.group(1)
    data = {}

    # Parse each line
    lines = resource_content.strip().split('\n')
    for line in lines:
        line = line.strip()
        if not line or line.startswith('#') or line.startswith('script'):
            continue

        # Match key = value
        match = re.match(r'^(\w+)\s*=\s*(.+)$', line)
        if match:
            key = match.group(1)
            value_str = match.group(2).strip()
            data[key] = parse_value(value_str)

    return data

def parse_value(value_str):
    """Parse a Godot value string to Python object."""
    value_str = value_str.strip()

    # Boolean
    if value_str == 'true':
        return True
    if value_str == 'false':
        return False

    # null/nil
    if value_str in ('null', 'nil'):
        return None

    # Color(r, g, b, a)
    color_match = re.match(r'Color\(\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)\s*\)', value_str)
    if color_match:
        return {
            "r": float(color_match.group(1)),
            "g": float(color_match.group(2)),
            "b": float(color_match.group(3)),
            "a": float(color_match.group(4))
        }

    # Vector2
    vec2_match = re.match(r'Vector2\(\s*([\d.-]+)\s*,\s*([\d.-]+)\s*\)', value_str)
    if vec2_match:
        return {"x": float(vec2_match.group(1)), "y": float(vec2_match.group(2))}

    # String
    if value_str.startswith('"') and value_str.endswith('"'):
        return value_str[1:-1]

    # Array
    if value_str.startswith('['):
        return parse_array(value_str)

    # Dictionary
    if value_str.startswith('{'):
        return parse_dict(value_str)

    # Integer
    if re.match(r'^-?\d+$', value_str):
        return int(value_str)

    # Float
    if re.match(r'^-?\d+\.\d*$', value_str):
        return float(value_str)

    return value_str

def parse_array(value_str):
    """Parse Godot array notation."""
    try:
        # Handle nested structures by using a more careful approach
        content = value_str[1:-1].strip()
        if not content:
            return []

        items = []
        depth = 0
        current = ""
        in_string = False

        for char in content:
            if char == '"' and (not current or current[-1] != '\\'):
                in_string = not in_string

            if not in_string:
                if char in '[{':
                    depth += 1
                elif char in ']}':
                    depth -= 1
                elif char == ',' and depth == 0:
                    items.append(parse_value(current.strip()))
                    current = ""
                    continue

            current += char

        if current.strip():
            items.append(parse_value(current.strip()))

        return items
    except:
        return []

def parse_dict(value_str):
    """Parse Godot dictionary notation."""
    try:
        content = value_str[1:-1].strip()
        if not content:
            return {}

        result = {}
        depth = 0
        current_key = None
        current_val = ""
        in_string = False
        parsing_key = True

        i = 0
        while i < len(content):
            char = content[i]

            if char == '"' and (i == 0 or content[i-1] != '\\'):
                in_string = not in_string

            if not in_string:
                if char in '[{':
                    depth += 1
                elif char in ']}':
                    depth -= 1
                elif char == ':' and depth == 0 and parsing_key:
                    current_key = current_val.strip().strip('"')
                    current_val = ""
                    parsing_key = False
                    i += 1
                    continue
                elif char == ',' and depth == 0:
                    if current_key is not None:
                        result[current_key] = parse_value(current_val.strip())
                    current_key = None
                    current_val = ""
                    parsing_key = True
                    i += 1
                    continue

            current_val += char
            i += 1

        if current_key is not None and current_val.strip():
            result[current_key] = parse_value(current_val.strip())

        return result
    except:
        return {}

def convert_monsters():
    """Convert all monster .tres files to JSON."""
    monsters_dir = GODOT_PROJECT / "data" / "monsters"
    output = []

    for tres_file in monsters_dir.glob("*.tres"):
        data = parse_tres_file(tres_file)
        if data:
            # Convert sprite paths from Godot to Unity format
            if 'sprite_path' in data:
                data['sprite_path'] = convert_path(data['sprite_path'])
            if 'portrait_path' in data:
                data['portrait_path'] = convert_path(data['portrait_path'])
            output.append(data)
            print(f"  Converted monster: {data.get('display_name', tres_file.stem)}")

    return output

def convert_skills():
    """Convert all skill .tres files to JSON."""
    skills_dir = GODOT_PROJECT / "data" / "skills"
    output = []

    # Main skills
    for tres_file in skills_dir.glob("*.tres"):
        data = parse_tres_file(tres_file)
        if data:
            if 'icon_path' in data:
                data['icon_path'] = convert_path(data['icon_path'])
            output.append(data)
            print(f"  Converted skill: {data.get('display_name', tres_file.stem)}")

    # Monster skills
    monster_skills_dir = skills_dir / "monsters"
    if monster_skills_dir.exists():
        for tres_file in monster_skills_dir.glob("*.tres"):
            data = parse_tres_file(tres_file)
            if data:
                data['is_monster_skill'] = True
                if 'icon_path' in data:
                    data['icon_path'] = convert_path(data['icon_path'])
                output.append(data)
                print(f"  Converted monster skill: {data.get('display_name', tres_file.stem)}")

    return output

def convert_heroes():
    """Convert all hero .tres files to JSON."""
    heroes_dir = GODOT_PROJECT / "data" / "heroes"
    output = []

    for tres_file in heroes_dir.glob("*.tres"):
        data = parse_tres_file(tres_file)
        if data:
            for path_key in ['sprite_path', 'portrait_path', 'battle_sprite_path']:
                if path_key in data:
                    data[path_key] = convert_path(data[path_key])
            output.append(data)
            print(f"  Converted hero: {data.get('display_name', tres_file.stem)}")

    # Also check nested heroes folder in skills
    heroes_dir2 = GODOT_PROJECT / "data" / "skills" / "heroes"
    if heroes_dir2.exists():
        for tres_file in heroes_dir2.glob("*.tres"):
            data = parse_tres_file(tres_file)
            if data:
                output.append(data)
                print(f"  Converted hero skill: {data.get('display_name', tres_file.stem)}")

    return output

def convert_items():
    """Convert all item .tres files to JSON."""
    items_dir = GODOT_PROJECT / "data" / "items"
    output = []

    for subdir in ['consumables', 'equipment']:
        subpath = items_dir / subdir
        if subpath.exists():
            for tres_file in subpath.glob("*.tres"):
                data = parse_tres_file(tres_file)
                if data:
                    data['item_category'] = subdir
                    if 'icon_path' in data:
                        data['icon_path'] = convert_path(data['icon_path'])
                    output.append(data)
                    print(f"  Converted item: {data.get('display_name', tres_file.stem)}")

    return output

def convert_path(godot_path):
    """Convert Godot res:// path to Unity-friendly path."""
    if not godot_path:
        return ""
    # Remove res:// prefix and convert to relative Unity path
    path = godot_path.replace("res://", "")
    # Convert to forward slashes
    path = path.replace("\\", "/")
    return path

def main():
    """Main conversion routine."""
    print("=" * 60)
    print("VEILBREAKERS DATA CONVERTER")
    print("Godot .tres -> Unity JSON")
    print("=" * 60)

    # Create output directory
    output_dir = UNITY_PROJECT / "Assets" / "Data"
    output_dir.mkdir(parents=True, exist_ok=True)

    # Convert monsters
    print("\n[MONSTERS]")
    monsters = convert_monsters()
    with open(output_dir / "monsters.json", 'w', encoding='utf-8') as f:
        json.dump(monsters, f, indent=2)
    print(f"  Total: {len(monsters)} monsters")

    # Convert skills
    print("\n[SKILLS]")
    skills = convert_skills()
    with open(output_dir / "skills.json", 'w', encoding='utf-8') as f:
        json.dump(skills, f, indent=2)
    print(f"  Total: {len(skills)} skills")

    # Convert heroes
    print("\n[HEROES]")
    heroes = convert_heroes()
    with open(output_dir / "heroes.json", 'w', encoding='utf-8') as f:
        json.dump(heroes, f, indent=2)
    print(f"  Total: {len(heroes)} heroes")

    # Convert items
    print("\n[ITEMS]")
    items = convert_items()
    with open(output_dir / "items.json", 'w', encoding='utf-8') as f:
        json.dump(items, f, indent=2)
    print(f"  Total: {len(items)} items")

    print("\n" + "=" * 60)
    print("CONVERSION COMPLETE!")
    print(f"Output directory: {output_dir}")
    print("=" * 60)

if __name__ == "__main__":
    main()
